### PR TITLE
Improve debug output of Time32/Time64 arrays

### LINKED ISSUE #5336: Debug output of Time32/Time64 arrays with invalid values has confusing nulls
**Describe the bug**

When debug printing a Time32/Time64 array with values outside of the expected range (between 0 seconds (inclusive) and 86,400 seconds (exclusive), adjusted based on milli/micro/nano), then it shows `null` for those values, which can be confused with actual null values.

**To Reproduce**

```rust
    #[test]
    fn test_pretty_format_time_32_second_invalid() {
        let array: Time32SecondArray = vec![
            Some(-1),
            Some(0),
            Some(86_399),
            Some(86_400),
            Some(86_401),
            None,
        ]
        .into();
        println!("{:?}", array);

        let schema = Arc::new(Schema::new(vec![Field::new(
            "f",
            array.data_type().clone(),
            true,
        )]));
        let batch = RecordBatch::try_new(schema, vec![Arc::new(array)]).unwrap();

        let table = pretty_format_batches(&[batch]).unwrap().to_string();
        println!("{}", table);
    }
```

Outputs:

```
PrimitiveArray<Time32(Second)>
[
  null,
  00:00:00,
  23:59:59,
  null,
  null,
  null,
]
+---------------------------------------------------------------------------+
| f                                                                         |
+---------------------------------------------------------------------------+
| ERROR: Cast error: Failed to convert -1 to temporal for Time32(Second)    |
| 00:00:00                                                                  |
| 23:59:59                                                                  |
| ERROR: Cast error: Failed to convert 86400 to temporal for Time32(Second) |
| ERROR: Cast error: Failed to convert 86401 to temporal for Time32(Second) |
|                                                                           |
+---------------------------------------------------------------------------+
```

**Expected behavior**

Either:

- Document this behaviour
- Print some other value for invalid values, like `<invalid>`, instead of just `null`

**Additional context**

Seems to be due to this code:

https://github.com/apache/arrow-rs/blob/8fff5e4a075c20619690c967e217163cd74e0656/arrow-array/src/array/primitive_array.rs#L1104-L1110

### PR DESCRIPTION
**Pull request by Monkwire3**
# Which issue does this PR close?

Closes #5336 .

# Rationale for this change
 This change makes the debug output of Time32/Time64 arrays more descriptive.

# What changes are included in this PR?
This PR changes the way Time32 and Time64 arrays are printed to debug. Previously, values that could not be cast to Time32 or Time64 had debug output values of `null`. With this change, unsuccessful casts are included as errors in the debug output.

# Are there any user-facing changes?
Yes, the debug output for Time32 and Time64 arrays will change.

### CODE REVIEW

File: arrow-array/src/array/primitive_array.rs [hunk 1]
```diff
@@ -1093,20 +1093,33 @@ where
...
                 let v = self.value(index).to_isize().unwrap() as i64;
                 match as_date::<T>(v) {
                     Some(date) => write!(f, "{date:?}"),
-                    None => write!(f, "null"),
+                    None => {
+                        let err = ArrowError::CastError(format!(
+                            "Failed to convert {} to temporal for {:?}",
+                            v, data_type
+                        ));
+                        write!(f, "{}", err)
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:1108]:
```suggestion
                        write!(
                            "Cast error: Failed to convert {} to temporal for {:?}",
                            v, data_type
                        );
```

File: arrow-array/src/array/primitive_array.rs [hunk 2]
```diff
@@ -1093,20 +1093,33 @@ where
...
                 let v = self.value(index).to_isize().unwrap() as i64;
                 match as_time::<T>(v) {
                     Some(time) => write!(f, "{time:?}"),
-                    None => write!(f, "null"),
+                    None => {
+                        let err = ArrowError::CastError(format!(
+                            "Failed to convert {} to temporal for {:?}",
+                            v, data_type
+                        ));
+                        write!(f, "{}", err)
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:1121]:
```suggestion
                        write!(
                            "Cast error: Failed to convert {} to temporal for {:?}",
                            v, data_type
                        );
```

**Review by tustvold — APPROVED** [review]
Just a minor suggestion, otherwise this looks good to me thank you

**Comment by Monkwire3** [comment]
Thanks, you are right, that looks better.

**Comment by tustvold** [comment]
I took the liberty of fixing the compilation errors in order to get this in

**Comment by Monkwire3** [comment]
Thank you!

**Pull request merged**